lib/recorder.h

```diff
@@ -0,0 +1,81 @@
+#pragma once
+
+#include <string>
+#include <fstream>
+#include <iostream>
+#include <sstream>
+#include <memory>
+
+namespace swss {
+
+class RecBase {
+public:
+    RecBase() {}
+    /* Setters */
+    void setRecord(bool record)  { enable_rec = record; }
+    void setRotate(bool rotate)  { log_rotate = rotate; }
+    void setLocation(const std::string& loc) { location = loc; }
+    void setFileName(const std::string& name) { filename = name; }
+    void setName(const std::string& name)  { m_name = name; }
+
+    /* getters */
+    bool isRecord()  { return enable_rec; }
+    bool isRotate()  { return log_rotate; }
+    std::string getLoc() { return location; }
+    std::string getFile() { return filename; }
+    std::string getName() { return m_name; }
+
+private:
+    bool enable_rec;
+    bool log_rotate;
+    std::string location;
+    std::string filename;
+    std::string m_name;
+};
+
+class RecWriter : public RecBase {
+public:
+    RecWriter() : RecBase() {}
+    virtual ~RecWriter();
+    void startRec(bool exit_if_failure);
+    void record(const std::string& val);
+
+protected:
+    void logfileReopen();
+
+private:
+    std::ofstream record_ofs;
+    std::string fname;
+};
+
+class SwSSRec : public RecWriter {
+public:
+    SwSSRec();
+};
+
+/* Record Handler for Response Publisher Class */
+class ResPubRec : public RecWriter {
+public:
+    ResPubRec();
+};
+
+class SaiRedisRec : public RecBase {
+public:
+    SaiRedisRec();
+};
```

Comment:
All 3 classes share the same interfaces. They fit better into 3 objects of class RecWriter.

Reply:
And that's why we have a Recorder interface.  

```
/* Interface to access recorder classes */
class Recorder {
public:
    static std::unique_ptr<SwSSRec> swss;
    static std::unique_ptr<SaiRedisRec> sairedis;
    static std::unique_ptr<ResPubRec> respub;
};
```

Reply:
The proposal "3 objects of class RecWriter" allow you to the same thing.

Reply:
Having 3 separate objects is flexible since we can easily extend the functionality for any one of these separately, if required.